packages/react-router-dom/index.tsx

```diff
@@ -334,12 +337,22 @@ export function useSearchParams(defaultInit?: URLSearchParamsInit) {
   }, [location.search]);
 
   let navigate = useNavigate();
+
   let setSearchParams = React.useCallback(
     (
-      nextInit: URLSearchParamsInit,
+      nextInit:
+        | URLSearchParamsInit
+        | ((params: URLSearchParams) => URLSearchParamsInit),
       navigateOptions?: { replace?: boolean; state?: State }
     ) => {
-      navigate('?' + createSearchParams(nextInit), navigateOptions);
+      const location = locationRef.current;
```

Comment:
As we don't want to pass `location` in useCallback dependency list, workaround is to store it on ref on render. When `setSearchParams` is called latest value will be available on ref. 

Imho should be safe in concurrent mode as 
- we don't mutate location, 
- setSearchParams can't be called in render

Reply:
Not relevant anymore, as changed to store only location search in ref